{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b8d88b7b-210b-41e4-ac56-701f926d205d",
   "metadata": {},
   "source": [
    "# Project BBT045\n",
    "Question:\n",
    "> How pharmaceutical pollution affect microbial community composition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b398bf9b-47a2-4bb2-9b6c-49607cbb6767",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load modules\n",
    "#??\n",
    "#matplotlib\n",
    "#seaborn?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e46a6853-bd0b-4df4-99ef-dc0d11e94451",
   "metadata": {},
   "source": [
    "# Temporary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "99dc2ca8-58a3-4453-b1d8-690de0e75386",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  - importlib-metadata=7.0.1=py310h06a4308_0\n",
      "  - importlib_metadata=7.0.1=hd3eb1b0_0\n",
      "  - metaphlan=4.0.6=pyhca03a8a_0\n"
     ]
    }
   ],
   "source": [
    "# Directories\n",
    "# groupDirectory = /cephyr/NOBACKUP/groups/bbt045_2024/project_group1/\n",
    "# resultsDirectory = /cephyr/NOBACKUP/groups/bbt045_2024/project_group1/results/\n",
    "# dataDirectory = /cephyr/NOBACKUP/groups/bbt045_2024/PROJECT_DATA\n",
    "\n",
    "\n",
    "# Checking stuff, remove later\n",
    "\n",
    "##Check that we have access to the files in the common directory:\n",
    "#with open(\"/cephyr/NOBACKUP/groups/bbt045_2024/project_group1/filelist.txt\") as f: # The with keyword automatically closes the file when you are done\n",
    "#    print(f.read())\n",
    "#    #ref https://stackoverflow.com/questions/18256363/how-do-i-print-the-content-of-a-txt-file-in-python\n",
    "\n",
    "#Search for packages in the softwarelist\n",
    "! grep \"meta\" < /cephyr/NOBACKUP/groups/bbt045_2024/ProjectSoftware/bbt045-projects.yml\n",
    "\n",
    "# Display the files in the data directory\n",
    "#!ls {dataDirectory}\n",
    "\n",
    "# Check variable directory names\n",
    "#!head {dataDirectory}/change_filenames_1.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2f7ee3e6-d8fe-4358-99d4-be1ed8dac44b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cephyr/NOBACKUP/groups/bbt045_2024/PROJECT_DATA/downstream_1.5km_2012\n",
      "/cephyr/NOBACKUP/groups/bbt045_2024/PROJECT_DATA/downstream_17.5km_2011\n",
      "/cephyr/NOBACKUP/groups/bbt045_2024/PROJECT_DATA/downstream_2.3km_2011\n",
      "/cephyr/NOBACKUP/groups/bbt045_2024/PROJECT_DATA/downstream_2.7km_2011\n",
      "/cephyr/NOBACKUP/groups/bbt045_2024/PROJECT_DATA/effluent_point_2011\n",
      "/cephyr/NOBACKUP/groups/bbt045_2024/PROJECT_DATA/upstream_1.9km_2011\n",
      "/cephyr/NOBACKUP/groups/bbt045_2024/PROJECT_DATA/upstream_150m_2012\n",
      "/cephyr/NOBACKUP/groups/bbt045_2024/PROJECT_DATA/upstream_2.2km_2011\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "DATADIR=/cephyr/NOBACKUP/groups/bbt045_2024/PROJECT_DATA;\n",
    "\n",
    "# Test av loop och filsökning med *:\n",
    "for file in `ls $DATADIR/*_1.fastq.gz  | grep -v \"_lake_\" | sed \"s/_1.fastq.gz//\"`\n",
    "do\n",
    "    echo $file\n",
    "    #outputName=`echo $file | sed \"s/\\/cephyr\\/NOBACKUP\\/groups\\/bbt045_2024\\/PROJECT_DATA\\///\"`\n",
    "    #echo $outputName\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00851db5-26e2-4d06-98c0-d64e2345a5df",
   "metadata": {},
   "source": [
    "# TO DO\n",
    "- Text hur vi valde saker\n",
    "- Text hur vi valde inställningar\n",
    "- Referenser: manualer?\n",
    "- Länk till script-filer? Fungerar det i Jupyter?\n",
    "- Uppdatera text med hur scriptet körs?\n",
    "- Ta bort onödiga kommentarer i kodrutorna\n",
    "- Renskriv batchfilernas kommentarer\n",
    "- Kopiera så att det finns två batchfiler för trim? 2012 vs 2011, se kommentarer\n",
    "- Fixa rescaled values\n",
    "- PCA plot i jupyter?\n",
    "- Andra plots i jupyter? Barplots? Boxplots?\n",
    "- Count species per sample?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a452085-4275-4d92-b73a-a3593a1c0b92",
   "metadata": {},
   "source": [
    "# Quality control and trimming"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f2f7b23-53c9-4db7-9ebe-934cd58d0819",
   "metadata": {},
   "source": [
    "## FastQC\n",
    "Do FastQC for all samples except those with \"lake\" in their name. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5cffa876-2fdc-4d11-a7f1-f52fec73fa8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%script false --no-raise-error\n",
    "for file in `ls /cephyr/NOBACKUP/groups/bbt045_2024/PROJECT_DATA/*.fastq.gz | grep -v \"_lake_\"`\n",
    "do\n",
    "    fastqc $file -o /cephyr/NOBACKUP/groups/bbt045_2024/project_group1/results/fastqc\n",
    "done\n",
    "\n",
    "#Tekniskt sett är det detta som körs:\n",
    "for file in `ls *.fastq.gz | grep -v \"_lake_\"`\n",
    "do\n",
    "    apptainer exec $CONTAINER_LOC fastqc $file -o $WORKING_TMP\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "659262ab-e7e5-4ed5-bec9-3ed749330fd1",
   "metadata": {},
   "source": [
    "## Bash script for trimmomatic"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56414c34-ecc2-44c1-a134-116c0888428c",
   "metadata": {},
   "source": [
    "The following sbatch script is to run trimmomatic on all samples except those with \"_lake_\" in their name\n",
    "\n",
    "NOTE:\n",
    "- Behövde använda -phred64 på 2011-filerna, gav error på dem annars när jag körde den. När ska köra 2012-filerna: byt ut år i `cp`och `-phred64` till `-phred33`\n",
    "\n",
    "Paired End Mode:\n",
    "ref: http://www.usadellab.org/cms/?page=trimmomatic\n",
    "> trimmomatic PE [-threads <threads>] [-phred33 | -phred64] [-trimlog <logFile>] <input 1> <input 2> <paired output 1> <unpaired output 1> <paired output 2> <unpaired output 2> <step 1> ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0a2de5c1-1d24-4483-902f-39a22a3a5cb9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%script false --no-raise-error\n",
    "# Extra för min förståelse, ta bort?\n",
    "WORKDIR=/cephyr/NOBACKUP/groups/bbt045_2024/project_group1/results/trimmomatic;\n",
    "WORKING_TMP=$TMPDIR/TRIM_TMP;\n",
    "\n",
    "mkdir $WORKING_TMP;\n",
    "cd $WORKING_TMP;\n",
    "\n",
    "cp /cephyr/NOBACKUP/groups/bbt045_2024/PROJECT_DATA/*2011*.fastq.gz $WORKING_TMP\n",
    "\n",
    "### Running Trimmomatic\n",
    "for file in `ls *_1.fastq.gz  | grep -v \"_lake_\" | sed \"s/_1.fastq.gz//\"`\n",
    "do\n",
    "    apptainer exec $CONTAINER_LOC trimmomatic PE -phred64 \\\n",
    "                    $file\\_1.fastq.gz $file\\_2.fastq.gz \\\n",
    "                    $file\\_1.trimmed.fastq.gz $file\\_1.un.trimmed.fastq.gz \\\n",
    "                    $file\\_2.trimmed.fastq.gz $file\\_2.un.trimmed.fastq.gz \\\n",
    "                    TOPHRED33 \\\n",
    "                    LEADING:30 TRAILING:30 \\\n",
    "                    MINLEN:60\n",
    "done\n",
    "\n",
    "\n",
    "# Extra, ta bort:\n",
    "    #question: does this output the files in $WORKING_TMP? Or do we need to add this before the output file names?\n",
    "            # or do we need to remove everything but the name and use that instead of $file as below?\n",
    "    #question/look at output: autodetect phred score? (run one of each and see? 33 first)\n",
    "    #question Paria/Alva: vilket score ska vi använda?\n",
    "    #question: Need to remove adapters? Nearly no present?\n",
    "    #question: trimmomatic phred score? TOPHRED64 if already PHRED64??\n",
    "\n",
    "    #ls $file\\_1.fastq.gz $file\\_2.fastq.gz\n",
    "    #apptainer exec $CONTAINER_LOC trim_galore --quality 20 --length 60 --cores 8 --paired -o /cephyr/NOBACKUP/groups/bbt045_2024/projectAliceRita5 $i\\_1.fastq.gz $i\\_2.fastq.gz\n",
    "\n",
    "# Johan brukar altid trimma adapters innan uppladdning, kanske därför vi inte ser några i fastqc-output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65b5fa4f-2be9-40d9-a947-43bcd1c76e28",
   "metadata": {},
   "source": [
    "# FastQC (again)\n",
    "Körde fastqc igen, fast denna gång på alla filer i project_group1/results/trimmomatic/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "680540de-51b4-45c1-875d-66d9956a21f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%script false --no-raise-error\n",
    "for file in `ls /cephyr/NOBACKUP/groups/bbt045_2024/PROJECT_DATA/*.fastq.gz | grep -v \"_lake_\"`\n",
    "do\n",
    "    fastqc $file -o /cephyr/NOBACKUP/groups/bbt045_2024/project_group1/results/fastqc\n",
    "done\n",
    "\n",
    "#Tekniskt sett är det detta som körs:\n",
    "for file in `ls *.fastq.gz | grep -v \"_lake_\"`\n",
    "do\n",
    "    apptainer exec $CONTAINER_LOC fastqc $file -o $WORKING_TMP\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24199954-802e-47b5-8f66-75009a5eefc7",
   "metadata": {},
   "source": [
    "# Metaphlan4\n",
    "\n",
    "Running metaphlan4\n",
    "\n",
    "Wiki: https://github.com/biobakery/biobakery/wiki/metaphlan4 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9759869e-7baf-4f83-a64a-3940fcbe7493",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%script false --no-raise-error\n",
    "#run the following in a terminal in a new tab in jupyter:\n",
    "metaphlan -h | less"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7229c561-7a17-4bc7-8f33-a9b0f927d612",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%script false --no-raise-error\n",
    "for file in `ls *1.trimmed.fastq.gz | sed \"s/_1.trimmed.fastq.gz//\"`\n",
    "do\n",
    "    apptainer exec $CONTAINER_LOC metaphlan $file\\_1.trimmed.fastq.gz,$file\\_2.trimmed.fastq.gz \\\n",
    "                                            --input_type fastq --nproc 16 --unclassified_estimation --bowtie2out $file.bowtie2.bz2 -o $file\\_profile.txt\\\n",
    "                                            \n",
    "done\n",
    "\n",
    "#From template from Vi:\n",
    "for i in SRS*.fasta.gz; do metaphlan $i --input_type fasta --nproc 4 > ${i%.fasta.gz}_profile.txt; done\n",
    "\n",
    "#todo Lägga till: ?\n",
    "#todo: PCA från pandas dataframe?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d83a37e-6fd0-404c-8d67-62b6e2cf42fc",
   "metadata": {},
   "source": [
    "## Visualizing metaphlan output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4183c7e8-50f7-4bd1-be14-7cbfd944f93e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%script false --no-raise-error\n",
    "# Merge the _profile tables for each sample into a single file\n",
    "!merge_metaphlan_tables.py /cephyr/NOBACKUP/groups/bbt045_2024/project_group1/results/metaphlan/*_profile.txt > merged_abundance_table.txt\n",
    "#ref: https://github.com/biobakery/biobakery/wiki/metaphlan4#16-merging-metaphlan-profiles\n",
    "\n",
    "# Isolate the species-resolved rows and then remove clutter \n",
    "!grep -E \"s__|clade_name|UNCLASSIFIED\" merged_abundance_table.txt |\\\n",
    "    grep -v \"t__\" |\\\n",
    "    sed \"s/^.*|s__//g\" |\\\n",
    "    sed \"s/SRS[0-9]*-//g\" \\\n",
    "    > merged_abundance_table_species.txt\n",
    "#ref https://github.com/biobakery/biobakery/wiki/metaphlan4#211-generate-a-species-only-abundance-table\n",
    "\n",
    "#remove s__ to keep it in the file, for specificity?\n",
    "#todo förbättra kommentarer ovan, se ref:s för mer info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c611f092-129e-480e-b8a6-304c3819df8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unclassified proportion: 92.6091\n",
      "\n",
      "Sum of species + unclassified:\n",
      " downstream_1.5km_2012     100.000003\n",
      "downstream_17.5km_2011    100.000003\n",
      "downstream_2.3km_2011     100.000005\n",
      "downstream_2.7km_2011     100.000000\n",
      "effluent_point_2011        99.999996\n",
      "upstream_1.9km_2011       100.000004\n",
      "upstream_150m_2012        100.000001\n",
      "upstream_2.2km_2011        99.999997\n",
      "dtype: float64\n",
      "\n",
      "Rescaled proportions sum, without unclassified:\n",
      " 100.00003513359341\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>downstream_1.5km_2012</th>\n",
       "      <th>downstream_17.5km_2011</th>\n",
       "      <th>downstream_2.3km_2011</th>\n",
       "      <th>downstream_2.7km_2011</th>\n",
       "      <th>effluent_point_2011</th>\n",
       "      <th>upstream_1.9km_2011</th>\n",
       "      <th>upstream_150m_2012</th>\n",
       "      <th>upstream_2.2km_2011</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clade_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>UNCLASSIFIED</th>\n",
       "      <td>92.609100</td>\n",
       "      <td>98.599950</td>\n",
       "      <td>97.334050</td>\n",
       "      <td>98.110410</td>\n",
       "      <td>96.755020</td>\n",
       "      <td>98.167200</td>\n",
       "      <td>88.277040</td>\n",
       "      <td>94.854580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Planktothricoides_sp_FACHB_1261</th>\n",
       "      <td>4.993868</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000721</td>\n",
       "      <td>0.017372</td>\n",
       "      <td>0.015698</td>\n",
       "      <td>0.033083</td>\n",
       "      <td>9.159325</td>\n",
       "      <td>0.015480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Aliarcobacter_cryaerophilus</th>\n",
       "      <td>0.518840</td>\n",
       "      <td>0.034597</td>\n",
       "      <td>0.039729</td>\n",
       "      <td>0.123864</td>\n",
       "      <td>0.312124</td>\n",
       "      <td>0.211597</td>\n",
       "      <td>0.562859</td>\n",
       "      <td>0.472879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGB41707_SGB58791</th>\n",
       "      <td>0.478770</td>\n",
       "      <td>0.001396</td>\n",
       "      <td>0.081815</td>\n",
       "      <td>0.019514</td>\n",
       "      <td>0.104623</td>\n",
       "      <td>0.002247</td>\n",
       "      <td>0.059514</td>\n",
       "      <td>0.290467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Methylomonas_sp_MP1</th>\n",
       "      <td>0.178137</td>\n",
       "      <td>0.021976</td>\n",
       "      <td>0.001380</td>\n",
       "      <td>0.001984</td>\n",
       "      <td>0.017945</td>\n",
       "      <td>0.000486</td>\n",
       "      <td>0.141465</td>\n",
       "      <td>0.039458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mesorhizobium_sp_YM1C_6_2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGB42747_SGB59953</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Anaeroarcus_burkinensis</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Aurantimicrobium_minutum</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zoogloea_ramigera</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000366</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>284 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 downstream_1.5km_2012  \\\n",
       "clade_name                                               \n",
       "UNCLASSIFIED                                 92.609100   \n",
       "Planktothricoides_sp_FACHB_1261               4.993868   \n",
       "Aliarcobacter_cryaerophilus                   0.518840   \n",
       "GGB41707_SGB58791                             0.478770   \n",
       "Methylomonas_sp_MP1                           0.178137   \n",
       "...                                                ...   \n",
       "Mesorhizobium_sp_YM1C_6_2                     0.000000   \n",
       "GGB42747_SGB59953                             0.000000   \n",
       "Anaeroarcus_burkinensis                       0.000000   \n",
       "Aurantimicrobium_minutum                      0.000000   \n",
       "Zoogloea_ramigera                             0.000000   \n",
       "\n",
       "                                 downstream_17.5km_2011  \\\n",
       "clade_name                                                \n",
       "UNCLASSIFIED                                  98.599950   \n",
       "Planktothricoides_sp_FACHB_1261                0.000000   \n",
       "Aliarcobacter_cryaerophilus                    0.034597   \n",
       "GGB41707_SGB58791                              0.001396   \n",
       "Methylomonas_sp_MP1                            0.021976   \n",
       "...                                                 ...   \n",
       "Mesorhizobium_sp_YM1C_6_2                      0.000000   \n",
       "GGB42747_SGB59953                              0.000000   \n",
       "Anaeroarcus_burkinensis                        0.000000   \n",
       "Aurantimicrobium_minutum                       0.000000   \n",
       "Zoogloea_ramigera                              0.000000   \n",
       "\n",
       "                                 downstream_2.3km_2011  downstream_2.7km_2011  \\\n",
       "clade_name                                                                      \n",
       "UNCLASSIFIED                                 97.334050              98.110410   \n",
       "Planktothricoides_sp_FACHB_1261               0.000721               0.017372   \n",
       "Aliarcobacter_cryaerophilus                   0.039729               0.123864   \n",
       "GGB41707_SGB58791                             0.081815               0.019514   \n",
       "Methylomonas_sp_MP1                           0.001380               0.001984   \n",
       "...                                                ...                    ...   \n",
       "Mesorhizobium_sp_YM1C_6_2                     0.000000               0.000000   \n",
       "GGB42747_SGB59953                             0.000000               0.000000   \n",
       "Anaeroarcus_burkinensis                       0.000000               0.000000   \n",
       "Aurantimicrobium_minutum                      0.000000               0.000000   \n",
       "Zoogloea_ramigera                             0.000000               0.000000   \n",
       "\n",
       "                                 effluent_point_2011  upstream_1.9km_2011  \\\n",
       "clade_name                                                                  \n",
       "UNCLASSIFIED                               96.755020            98.167200   \n",
       "Planktothricoides_sp_FACHB_1261             0.015698             0.033083   \n",
       "Aliarcobacter_cryaerophilus                 0.312124             0.211597   \n",
       "GGB41707_SGB58791                           0.104623             0.002247   \n",
       "Methylomonas_sp_MP1                         0.017945             0.000486   \n",
       "...                                              ...                  ...   \n",
       "Mesorhizobium_sp_YM1C_6_2                   0.000000             0.000000   \n",
       "GGB42747_SGB59953                           0.000000             0.000000   \n",
       "Anaeroarcus_burkinensis                     0.000000             0.000000   \n",
       "Aurantimicrobium_minutum                    0.000000             0.000000   \n",
       "Zoogloea_ramigera                           0.000000             0.000000   \n",
       "\n",
       "                                 upstream_150m_2012  upstream_2.2km_2011  \n",
       "clade_name                                                                \n",
       "UNCLASSIFIED                              88.277040            94.854580  \n",
       "Planktothricoides_sp_FACHB_1261            9.159325             0.015480  \n",
       "Aliarcobacter_cryaerophilus                0.562859             0.472879  \n",
       "GGB41707_SGB58791                          0.059514             0.290467  \n",
       "Methylomonas_sp_MP1                        0.141465             0.039458  \n",
       "...                                             ...                  ...  \n",
       "Mesorhizobium_sp_YM1C_6_2                  0.000000             0.000772  \n",
       "GGB42747_SGB59953                          0.000000             0.000663  \n",
       "Anaeroarcus_burkinensis                    0.000000             0.000588  \n",
       "Aurantimicrobium_minutum                   0.000000             0.000381  \n",
       "Zoogloea_ramigera                          0.000000             0.000366  \n",
       "\n",
       "[284 rows x 8 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "species_file = \"merged_abundance_table_species.txt\"\n",
    "# Tab separated, first row column names, first column row names\n",
    "species_df = pd.read_csv(species_file, sep='\\t', header=0, index_col=0)\n",
    "\n",
    "unclassified__proportion = species_df.iloc[0,0]\n",
    "print(\"Unclassified proportion:\", unclassified__proportion)\n",
    "\n",
    "species_df.head(25)\n",
    "print(\"\\nSum of species + unclassified:\\n\", species_df.sum())\n",
    "\n",
    "# Rescale all values:\n",
    "test = (100-unclassified__proportion)/100\n",
    "test_df = species_df/test\n",
    "test_proportions = test_df.iloc[range(1, len(test_df-1)), 0].sum()\n",
    "print(\"\\nRescaled proportions sum, without unclassified:\\n\", test_proportions)\n",
    "\n",
    "#test_df #View the rescaled proportions (comment out the other one)\n",
    "species_df #View the species dataframe\n",
    "\n",
    "#todo MERGE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "91d5c52b-7e52-4024-b13c-2e10e5f1125a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%script false --no-raise-error\n",
    "#todo EXAMPLE; REMOVE\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "species = species_df.index # y axis\n",
    "samples = species_df.columns # x axis\n",
    "proportion = np.array(species_df) # value in cell\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [12, 12]\n",
    "fig, ax = plt.subplots()\n",
    "im = ax.imshow(proportion)\n",
    "\n",
    "# Show all ticks and label them with the respective list entries\n",
    "ax.set_xticks(np.arange(len(samples)), labels=samples)\n",
    "ax.set_yticks(np.arange(len(species)), labels=species)\n",
    "\n",
    "# Rotate the tick labels and set their alignment.\n",
    "plt.setp(ax.get_xticklabels(), rotation=45, ha=\"right\",\n",
    "         rotation_mode=\"anchor\")\n",
    "\n",
    "# Loop over data dimensions and create text annotations.\n",
    "for i in range(len(species)):\n",
    "    for j in range(len(samples)):\n",
    "        text = ax.text(j, i, proportion[i, j].round(2),\n",
    "                       ha=\"center\", va=\"center\", color=\"w\")\n",
    "\n",
    "ax.set_title(\"Proportion of species per sample\")\n",
    "fig.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "species_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5b451de7-4e73-4c2c-9f2b-23389ba5a5e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ref: https://matplotlib.org/stable/gallery/images_contours_and_fields/image_annotated_heatmap.html\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "import matplotlib\n",
    "\n",
    "\n",
    "def heatmap(data, row_labels, col_labels, ax=None,\n",
    "            cbar_kw=None, cbarlabel=\"\", **kwargs):\n",
    "    \"\"\"\n",
    "    Create a heatmap from a numpy array and two lists of labels.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    data\n",
    "        A 2D numpy array of shape (M, N).\n",
    "    row_labels\n",
    "        A list or array of length M with the labels for the rows.\n",
    "    col_labels\n",
    "        A list or array of length N with the labels for the columns.\n",
    "    ax\n",
    "        A `matplotlib.axes.Axes` instance to which the heatmap is plotted.  If\n",
    "        not provided, use current axes or create a new one.  Optional.\n",
    "    cbar_kw\n",
    "        A dictionary with arguments to `matplotlib.Figure.colorbar`.  Optional.\n",
    "    cbarlabel\n",
    "        The label for the colorbar.  Optional.\n",
    "    **kwargs\n",
    "        All other arguments are forwarded to `imshow`.\n",
    "    \"\"\"\n",
    "\n",
    "    if ax is None:\n",
    "        ax = plt.gca()\n",
    "\n",
    "    if cbar_kw is None:\n",
    "        cbar_kw = {}\n",
    "\n",
    "    # Plot the heatmap\n",
    "    im = ax.imshow(data, **kwargs)\n",
    "\n",
    "    # Create colorbar\n",
    "    cbar = ax.figure.colorbar(im, ax=ax, **cbar_kw)\n",
    "    cbar.ax.set_ylabel(cbarlabel, rotation=-90, va=\"bottom\")\n",
    "\n",
    "    # Show all ticks and label them with the respective list entries.\n",
    "    ax.set_xticks(np.arange(data.shape[1]), labels=col_labels)\n",
    "    ax.set_yticks(np.arange(data.shape[0]), labels=row_labels)\n",
    "\n",
    "    # Let the horizontal axes labeling appear on top.\n",
    "    ax.tick_params(top=True, bottom=False,\n",
    "                   labeltop=True, labelbottom=False)\n",
    "\n",
    "    # Rotate the tick labels and set their alignment.\n",
    "    plt.setp(ax.get_xticklabels(), rotation=-30, ha=\"right\",\n",
    "             rotation_mode=\"anchor\")\n",
    "\n",
    "    # Turn spines off and create white grid.\n",
    "    ax.spines[:].set_visible(False)\n",
    "\n",
    "    ax.set_xticks(np.arange(data.shape[1]+1)-.5, minor=True)\n",
    "    ax.set_yticks(np.arange(data.shape[0]+1)-.5, minor=True)\n",
    "    ax.grid(which=\"minor\", color=\"w\", linestyle='-', linewidth=3)\n",
    "    ax.tick_params(which=\"minor\", bottom=False, left=False)\n",
    "\n",
    "    return im, cbar\n",
    "\n",
    "\n",
    "def annotate_heatmap(im, data=None, valfmt=\"{x:.2f}\",\n",
    "                     textcolors=(\"black\", \"white\"),\n",
    "                     threshold=None, **textkw):\n",
    "    \"\"\"\n",
    "    A function to annotate a heatmap.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    im\n",
    "        The AxesImage to be labeled.\n",
    "    data\n",
    "        Data used to annotate.  If None, the image's data is used.  Optional.\n",
    "    valfmt\n",
    "        The format of the annotations inside the heatmap.  This should either\n",
    "        use the string format method, e.g. \"$ {x:.2f}\", or be a\n",
    "        `matplotlib.ticker.Formatter`.  Optional.\n",
    "    textcolors\n",
    "        A pair of colors.  The first is used for values below a threshold,\n",
    "        the second for those above.  Optional.\n",
    "    threshold\n",
    "        Value in data units according to which the colors from textcolors are\n",
    "        applied.  If None (the default) uses the middle of the colormap as\n",
    "        separation.  Optional.\n",
    "    **kwargs\n",
    "        All other arguments are forwarded to each call to `text` used to create\n",
    "        the text labels.\n",
    "    \"\"\"\n",
    "\n",
    "    if not isinstance(data, (list, np.ndarray)):\n",
    "        data = im.get_array()\n",
    "\n",
    "    # Normalize the threshold to the images color range.\n",
    "    if threshold is not None:\n",
    "        threshold = im.norm(threshold)\n",
    "    else:\n",
    "        threshold = im.norm(data.max())/2.\n",
    "\n",
    "    # Set default alignment to center, but allow it to be\n",
    "    # overwritten by textkw.\n",
    "    kw = dict(horizontalalignment=\"center\",\n",
    "              verticalalignment=\"center\")\n",
    "    kw.update(textkw)\n",
    "\n",
    "    # Get the formatter in case a string is supplied\n",
    "    if isinstance(valfmt, str):\n",
    "        valfmt = matplotlib.ticker.StrMethodFormatter(valfmt)\n",
    "\n",
    "    # Loop over the data and create a `Text` for each \"pixel\".\n",
    "    # Change the text's color depending on the data.\n",
    "    texts = []\n",
    "    for i in range(data.shape[0]):\n",
    "        for j in range(data.shape[1]):\n",
    "            kw.update(color=textcolors[int(im.norm(data[i, j]) > threshold)])\n",
    "            text = im.axes.text(j, i, valfmt(data[i, j], None), **kw)\n",
    "            texts.append(text)\n",
    "\n",
    "    return texts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29f3fac2-21e3-47e1-af2f-ea4c84f70668",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the arrays for the functions to use\n",
    "species = species_df.index # y axis\n",
    "samples = species_df.columns # x axis\n",
    "proportion = np.array(species_df) # value in cell\n",
    "\n",
    "#ref https://stackoverflow.com/questions/36367986/how-to-make-inline-plots-in-jupyter-notebook-larger\n",
    "plt.rcParams['figure.figsize'] = [12, 120]\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "im, cbar = heatmap(proportion, species, samples, ax=ax,\n",
    "                   cmap=\"YlGn\", cbarlabel=\"Proportion (%)\")\n",
    "texts = annotate_heatmap(im, valfmt=\"{x:.1f}\") #todo Where the values are rounded \n",
    "\n",
    "fig.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d17000de-fc3d-4be4-8892-22f15a14a1a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRS014459-Stool_profile.txt\n",
      "SRS014464-Anterior_nares_profile.txt\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "for file in `ls SRS*.txt`\n",
    "do\n",
    "echo $file\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9b330f95-8aaf-46c1-8b76-865e57a5adb7",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'all_profile_files' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[15], line 4\u001b[0m\n\u001b[1;32m      2\u001b[0m get_ipython()\u001b[38;5;241m.\u001b[39msystem(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m$all_profile_files\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m#{species_files}\u001b[39;00m\n\u001b[0;32m----> 4\u001b[0m \u001b[43mall_profile_files\u001b[49m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'all_profile_files' is not defined"
     ]
    }
   ],
   "source": [
    "!all_profile_files=`ls *_profile.txt`\n",
    "!$all_profile_files\n",
    "#{species_files}\n",
    "all_profile_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0d6e382-987e-4285-8de0-6b2593f7bd12",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PCA\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f259c942-caa1-48f8-8256-6611d117eff1",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "206f30b6-d916-4c8e-9f16-f454d5b49d56",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.pairplot(species_df, hue='upstream_2.2km_2011')\n",
    "#seaborn.pairplot(data, *, hue=None, hue_order=None, palette=None, vars=None, x_vars=None, y_vars=None, kind='scatter', diag_kind='auto', markers=None, height=2.5, aspect=1, corner=False, dropna=False, plot_kws=None, diag_kws=None, grid_kws=None, size=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c159305-6d06-4adf-81fb-91edb2aa03f1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
